File: backend/src/main/java/com/caixy/adminSystem/controller/RegistrationTaskLessonController.java
package com.caixy.adminSystem.controller;

import com.baomidou.mybatisplus.extension.plugins.pagination.Page;
import com.caixy.adminSystem.annotation.AuthCheck;
import com.caixy.adminSystem.common.BaseResponse;
import com.caixy.adminSystem.common.DeleteRequest;
import com.caixy.adminSystem.common.ErrorCode;
import com.caixy.adminSystem.common.ResultUtils;
import com.caixy.adminSystem.constant.UserConstant;
import com.caixy.adminSystem.exception.BusinessException;
import com.caixy.adminSystem.exception.ThrowUtils;
import com.caixy.adminSystem.model.dto.registrationTaskLesson.*;
import com.caixy.adminSystem.model.entity.RegistrationTaskLesson;
import com.caixy.adminSystem.model.entity.User;
import com.caixy.adminSystem.model.vo.registrationTaskLesson.HasRegistrationTaskVO;
import com.caixy.adminSystem.model.vo.registrationTaskLesson.RegistrationTaskLessonVO;
import com.caixy.adminSystem.service.RegistrationTaskLessonService;
import com.caixy.adminSystem.service.UserService;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.BeanUtils;
import org.springframework.web.bind.annotation.*;

import javax.annotation.Resource;
import javax.servlet.http.HttpServletRequest;
import javax.validation.Valid;
import java.util.List;

/**
 * 登分任务课程信息接口
 */
@RestController
@RequestMapping("/registrationTaskLesson")
@Slf4j
public class RegistrationTaskLessonController
{

    @Resource
    private RegistrationTaskLessonService registrationTaskLessonService;

    @Resource
    private UserService userService;

    // region 增删改查

    /**
     * 创建登分任务课程信息
     *
     * @param registrationTaskLessonAddRequest
     * @param request
     * @return
     */
    @PostMapping("/add")
    public BaseResponse<Long> addRegistrationTaskLesson(@RequestBody RegistrationTaskLessonAddRequest registrationTaskLessonAddRequest, HttpServletRequest request)
    {
        ThrowUtils.throwIf(registrationTaskLessonAddRequest == null, ErrorCode.PARAMS_ERROR);
        // todo 在此处将实体类和 DTO 进行转换
        RegistrationTaskLesson registrationTaskLesson = new RegistrationTaskLesson();
        BeanUtils.copyProperties(registrationTaskLessonAddRequest, registrationTaskLesson);
        // 数据校验
        registrationTaskLessonService.validRegistrationTaskLesson(registrationTaskLesson, true);
        // todo 填充默认值
        // 写入数据库
        boolean result = registrationTaskLessonService.save(registrationTaskLesson);
        ThrowUtils.throwIf(!result, ErrorCode.OPERATION_ERROR);
        // 返回新写入的数据 id
        long newRegistrationTaskLessonId = registrationTaskLesson.getId();
        return ResultUtils.success(newRegistrationTaskLessonId);
    }

    /**
     * 删除登分任务课程信息
     *
     * @param deleteRequest
     * @param request
     * @return
     */
    @PostMapping("/delete")
    public BaseResponse<Boolean> deleteRegistrationTaskLesson(@RequestBody DeleteRequest deleteRequest, HttpServletRequest request)
    {
        if (deleteRequest == null || deleteRequest.getId() <= 0)
        {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        User user = userService.getLoginUser(request);
        long id = deleteRequest.getId();
        // 判断是否存在
        RegistrationTaskLesson oldRegistrationTaskLesson = registrationTaskLessonService.getById(id);
        ThrowUtils.throwIf(oldRegistrationTaskLesson == null, ErrorCode.NOT_FOUND_ERROR);
        // 仅本人或管理员可删除
        checkIsSelfOrAdmin(oldRegistrationTaskLesson, user);
        // 操作数据库
        boolean result = registrationTaskLessonService.removeById(id);
        ThrowUtils.throwIf(!result, ErrorCode.OPERATION_ERROR);
        return ResultUtils.success(true);
    }

    /**
     * 更新登分任务课程信息（仅管理员可用）
     *
     * @param registrationTaskLessonUpdateRequest
     * @return
     */
    @PostMapping("/update")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Boolean> updateRegistrationTaskLesson(@RequestBody RegistrationTaskLessonUpdateRequest registrationTaskLessonUpdateRequest)
    {
        if (registrationTaskLessonUpdateRequest == null || registrationTaskLessonUpdateRequest.getId() <= 0)
        {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        // todo 在此处将实体类和 DTO 进行转换
        RegistrationTaskLesson registrationTaskLesson = new RegistrationTaskLesson();
        BeanUtils.copyProperties(registrationTaskLessonUpdateRequest, registrationTaskLesson);
        // 数据校验
        registrationTaskLessonService.validRegistrationTaskLesson(registrationTaskLesson, false);
        // 判断是否存在
        long id = registrationTaskLessonUpdateRequest.getId();
        RegistrationTaskLesson oldRegistrationTaskLesson = registrationTaskLessonService.getById(id);
        ThrowUtils.throwIf(oldRegistrationTaskLesson == null, ErrorCode.NOT_FOUND_ERROR);
        // 操作数据库
        boolean result = registrationTaskLessonService.updateById(registrationTaskLesson);
        ThrowUtils.throwIf(!result, ErrorCode.OPERATION_ERROR);
        return ResultUtils.success(true);
    }


    /**
     * 分页获取登分任务课程信息列表（仅管理员可用）
     *
     * @param registrationTaskLessonQueryRequest
     * @return
     */
    @PostMapping("/list/page")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Page<RegistrationTaskLesson>> listRegistrationTaskLessonByPage(@RequestBody RegistrationTaskLessonQueryRequest registrationTaskLessonQueryRequest)
    {
        long current = registrationTaskLessonQueryRequest.getCurrent();
        long size = registrationTaskLessonQueryRequest.getPageSize();
        // 查询数据库
        Page<RegistrationTaskLesson> registrationTaskLessonPage = registrationTaskLessonService.page(new Page<>(current, size), registrationTaskLessonService.getQueryWrapper(registrationTaskLessonQueryRequest));
        return ResultUtils.success(registrationTaskLessonPage);
    }

    /**
     * 分页获取登分任务课程信息列表（封装类）
     *
     * @param registrationTaskLessonQueryRequest
     * @param request
     * @return
     */
    @PostMapping("/list/page/vo")
    public BaseResponse<Page<RegistrationTaskLessonVO>> listRegistrationTaskLessonVOByPage(@RequestBody RegistrationTaskLessonQueryRequest registrationTaskLessonQueryRequest, HttpServletRequest request)
    {
        long current = registrationTaskLessonQueryRequest.getCurrent();
        long size = registrationTaskLessonQueryRequest.getPageSize();
        // 限制爬虫
        ThrowUtils.throwIf(size > 20, ErrorCode.PARAMS_ERROR);
        // 查询数据库
        Page<RegistrationTaskLesson> registrationTaskLessonPage = registrationTaskLessonService.page(new Page<>(current, size), registrationTaskLessonService.getQueryWrapper(registrationTaskLessonQueryRequest));
        // 获取封装类
        return ResultUtils.success(registrationTaskLessonService.getRegistrationTaskLessonVOPage(registrationTaskLessonPage, request));
    }

    /**
     * 分页获取当前登录用户创建的登分任务课程信息列表
     *
     * @param registrationTaskLessonQueryRequest
     * @param request
     * @return
     */
    @PostMapping("/my/list/page/vo")
    public BaseResponse<Page<RegistrationTaskLessonVO>> listMyRegistrationTaskLessonVOByPage(@RequestBody RegistrationTaskLessonQueryRequest registrationTaskLessonQueryRequest, HttpServletRequest request)
    {
        ThrowUtils.throwIf(registrationTaskLessonQueryRequest == null, ErrorCode.PARAMS_ERROR);
        // 补充查询条件，只查询当前登录用户的数据
        User loginUser = userService.getLoginUser(request);
        registrationTaskLessonQueryRequest.setUserId(loginUser.getId());
        long current = registrationTaskLessonQueryRequest.getCurrent();
        long size = registrationTaskLessonQueryRequest.getPageSize();
        // 限制爬虫
        ThrowUtils.throwIf(size > 20, ErrorCode.PARAMS_ERROR);
        // 查询数据库
        Page<RegistrationTaskLesson> registrationTaskLessonPage = registrationTaskLessonService.page(new Page<>(current, size), registrationTaskLessonService.getQueryWrapper(registrationTaskLessonQueryRequest));
        // 获取封装类
        return ResultUtils.success(registrationTaskLessonService.getRegistrationTaskLessonVOPage(registrationTaskLessonPage, request));
    }

    /**
     * 编辑登分任务课程信息（给用户使用）
     *
     * @param registrationTaskLessonEditRequest
     * @param request
     * @return
     */
    @PostMapping("/edit")
    public BaseResponse<Boolean> editRegistrationTaskLesson(@RequestBody RegistrationTaskLessonEditRequest registrationTaskLessonEditRequest, HttpServletRequest request)
    {
        if (registrationTaskLessonEditRequest == null || registrationTaskLessonEditRequest.getId() <= 0)
        {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        // todo 在此处将实体类和 DTO 进行转换
        RegistrationTaskLesson registrationTaskLesson = new RegistrationTaskLesson();
        BeanUtils.copyProperties(registrationTaskLessonEditRequest, registrationTaskLesson);
        // 数据校验
        registrationTaskLessonService.validRegistrationTaskLesson(registrationTaskLesson, false);
        User loginUser = userService.getLoginUser(request);
        // 判断是否存在
        long id = registrationTaskLessonEditRequest.getId();
        RegistrationTaskLesson oldRegistrationTaskLesson = registrationTaskLessonService.getById(id);
        ThrowUtils.throwIf(oldRegistrationTaskLesson == null, ErrorCode.NOT_FOUND_ERROR);
        // 仅本人或管理员可编辑
        checkIsSelfOrAdmin(oldRegistrationTaskLesson, loginUser);
        // 操作数据库
        boolean result = registrationTaskLessonService.updateById(registrationTaskLesson);
        ThrowUtils.throwIf(!result, ErrorCode.OPERATION_ERROR);
        return ResultUtils.success(true);
    }

    @PostMapping("/get/hasTasks")
    public BaseResponse<List<HasRegistrationTaskVO>> getHasTasksBySubjectIdAndCourseTaskIds(@RequestBody @Valid HasRegistrationTaskRequest hasRegistrationTaskRequest)
    {
        return ResultUtils.success(registrationTaskLessonService.getHasTasksBySubjectIdAndCourseTaskIds(hasRegistrationTaskRequest.getSubjectId(), hasRegistrationTaskRequest.getCourseTaskIds()));
    }

    /**
     * 获取等分任务课程信息
     *
     * @author CAIXYPROMISE
     * @version 1.0
     * @version 2025/1/27 1:25
     */
    @GetMapping("/get/registrationTaskLessonVO/{id}")
    public BaseResponse<List<RegistrationTaskLessonVO>> listRegistrationTaskLessonVOByPage(@PathVariable("id") Long id)
    {
        return ResultUtils.success(registrationTaskLessonService.getRegistrationLessonByTaskId(id));
    }

    @PostMapping("/update/publication")
    public BaseResponse<Boolean> updatePublication(@RequestBody @Valid RegistrationTaskLessonPublicationRequest registrationTaskLessonPublicationRequest)
    {
        return ResultUtils.success(registrationTaskLessonService.publicationGradeInfo(registrationTaskLessonPublicationRequest));
    }

    // endregion

    private void checkIsSelfOrAdmin(RegistrationTaskLesson registrationTaskLesson, User loginUser)
    {

    }
}

File: frontend/src/services/backend/registrationTaskLessonController.ts
// @ts-ignore
/* eslint-disable */
import { request } from '@umijs/max';

/** addRegistrationTaskLesson POST /api/registrationTaskLesson/add */
export async function addRegistrationTaskLessonUsingPost1(
  body: API.RegistrationTaskLessonAddRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseLong_>('/api/registrationTaskLesson/add', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** deleteRegistrationTaskLesson POST /api/registrationTaskLesson/delete */
export async function deleteRegistrationTaskLessonUsingPost1(
  body: API.DeleteRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseBoolean_>('/api/registrationTaskLesson/delete', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** editRegistrationTaskLesson POST /api/registrationTaskLesson/edit */
export async function editRegistrationTaskLessonUsingPost1(
  body: API.RegistrationTaskLessonEditRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseBoolean_>('/api/registrationTaskLesson/edit', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** getHasTasksBySubjectIdAndCourseTaskIds POST /api/registrationTaskLesson/get/hasTasks */
export async function getHasTasksBySubjectIdAndCourseTaskIdsUsingPost1(
  body: API.HasRegistrationTaskRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseListHasRegistrationTaskVO_>(
    '/api/registrationTaskLesson/get/hasTasks',
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      data: body,
      ...(options || {}),
    },
  );
}

/** listRegistrationTaskLessonVOByPage GET /api/registrationTaskLesson/get/registrationTaskLessonVO/${param0} */
export async function listRegistrationTaskLessonVoByPageUsingGet1(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.listRegistrationTaskLessonVOByPageUsingGET1Params,
  options?: { [key: string]: any },
) {
  const { id: param0, ...queryParams } = params;
  return request<API.BaseResponseListRegistrationTaskLessonVO_>(
    `/api/registrationTaskLesson/get/registrationTaskLessonVO/${param0}`,
    {
      method: 'GET',
      params: { ...queryParams },
      ...(options || {}),
    },
  );
}

/** listRegistrationTaskLessonByPage POST /api/registrationTaskLesson/list/page */
export async function listRegistrationTaskLessonByPageUsingPost1(
  body: API.RegistrationTaskLessonQueryRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponsePageRegistrationTaskLesson_>(
    '/api/registrationTaskLesson/list/page',
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      data: body,
      ...(options || {}),
    },
  );
}

/** listRegistrationTaskLessonVOByPage POST /api/registrationTaskLesson/list/page/vo */
export async function listRegistrationTaskLessonVoByPageUsingPost1(
  body: API.RegistrationTaskLessonQueryRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponsePageRegistrationTaskLessonVO_>(
    '/api/registrationTaskLesson/list/page/vo',
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      data: body,
      ...(options || {}),
    },
  );
}

/** listMyRegistrationTaskLessonVOByPage POST /api/registrationTaskLesson/my/list/page/vo */
export async function listMyRegistrationTaskLessonVoByPageUsingPost1(
  body: API.RegistrationTaskLessonQueryRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponsePageRegistrationTaskLessonVO_>(
    '/api/registrationTaskLesson/my/list/page/vo',
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      data: body,
      ...(options || {}),
    },
  );
}

/** updateRegistrationTaskLesson POST /api/registrationTaskLesson/update */
export async function updateRegistrationTaskLessonUsingPost1(
  body: API.RegistrationTaskLessonUpdateRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseBoolean_>('/api/registrationTaskLesson/update', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** updatePublication POST /api/registrationTaskLesson/update/publication */
export async function updatePublicationUsingPost1(
  body: API.RegistrationTaskLessonPublicationRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseBoolean_>('/api/registrationTaskLesson/update/publication', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}
